### ISSUE TITLE
Testing???!!!

### ISSUE THREAD

**Issue by Jester565** [Jester565/SpyNinjaGame#25]
The life count, ninja attack, and win state system are not working...

**Comment by jtara1** [comment]
The life count works for me. Are you sure you `git pull` and `git checkout master` or maybe 
`git checkout origin/master`?

**I'm pretty sure the only feature not implemented / not working is the death of a ninja who has been shot.**

edit: I tested the three things you mentioned, they work as expected when I play the game.

 Here's game prints from me winning the game

```
A room is blocking your vision ahead
[ ][ ][ ][ ][ ][ ][N][ ][ ]   Lives: 2
[ ][R][ ][ ][R][N][ ][R][B]   Ammo: 1
[ ][ ][ ][N][ ][ ][I][ ][ ]   Invincible Turns: 0
[ ][ ][ ][ ][ ][r][ ][ ][ ]   Radar: Disabled
[ ][R][ ][ ][R][ ][ ][R][ ]
[ ][N][ ][ ][ ][ ][ ][ ][ ]
[ ][ ][N][ ][ ][ ][ ][S][ ]
[ ][R][ ][ ][R][ ][ ][b][ ]
[ ][ ][ ][ ][ ][ ][N][ ][ ]

Enter a direction to move or another command
W: Up | A: Left | S: Down | D: Right
1: Shoot | 2: Debug | 3: More Options
s
Move Status: You found the briefcase!
You Win!
[ ][ ][ ][ ][ ][N][ ][ ][ ]   Lives: 2
[ ][R][ ][ ][R][ ][N][R][B]   Ammo: 1
[ ][ ][ ][ ][ ][ ][I][ ][ ]   Invincible Turns: 0
[ ][ ][ ][N][ ][r][ ][ ][ ]   Radar: Disabled
[ ][R][ ][ ][R][ ][ ][R][ ]
[N][ ][ ][ ][ ][ ][ ][ ][ ]
[ ][ ][ ][ ][ ][ ][ ][S][ ]
[ ][R][N][ ][R][ ][ ][b][ ]
[ ][ ][ ][ ][ ][N][ ][ ][ ]

Select an option:
1: Start New Game
2: Load Game
3: Exit
```

**Comment by jtara1** [comment]
@walkerzheng @frfabianrp @jrcastron @kjasonk 
Could we get some more testers?

If you guys aren't too busy,

please play the game and test out at least a few features.

Alex says he'll work on bug fixing, but I'm not seeing the same bugs he is.

Here's a list of things you can test:

* [x] player look
* [x] player move
* [x] player step on bullet
* [x] player step on invincibility
* [x] player step on radar
* [x] player step in range of a ninja and loses a life
* [x] player shoot in direction (does not remove ninja from game if ninja shot at the moment)
* [x] player attempts to move out of bounds 
* [x] player attempts to move into room on wrong side
* [x] player searches room (from north side) for briefcase
* [x] player wins game if briefcase is found
* [x] player loses game if he dies with 0 lives
* [x] ninjas each move in random direction
* [x] enable & disable debug mode
* [x] More Options options works as expected (don't need to worry about save/load features for tomorrow though)

Feel free to edit my comment if you want to copy & paste this list in your own comment.

**Comment by Jester565** [comment]
Don't do anything with the repository with the issues I put here.  I already fixed them.  I'm going to put the issues I found here.

1. Enable debug mode and disabling it in the same turn gets rid of visible markers. (fixed)
2. A Ninja killed me one turn after I spawned in. That shouldn't be possible.

**Comment by jtara1** [comment]
> A Ninja killed me one turn after I spawned in. That shouldn't be possible.

I'm pretty sure each ninja does 3 things per turn and that might be the cause of this issue:
1. stab player if adjacent
2. move in random direction
3. ~~stab player if adjacent~~ (fixed)

**Comment by Jester565** [comment]
Yeah.  I saw that.  I'll get to it in a second.

**Comment by YaysonKwok** [comment]
~~Life counter seems to work sometimes but not others. Here I went from 3 to 1 life after dying once.~~ (fixed)
```
Clear
[N][ ][ ][ ][ ][ ][ ][ ][ ]   Lives: 3
[r][R][ ][ ][R][ ][ ][R][ ]   Ammo: 1
[ ][ ][ ][ ][ ][ ][ ][ ][ ]   Invincible Turns: 1
[N][ ][ ][ ][ ][ ][ ][ ][N]   Radar: Disabled
[ ][R][ ][ ][b][N][ ][R][ ]
[ ][ ][ ][N][ ][ ][ ][ ][ ]
[ ][ ][S][ ][ ][ ][B][ ][ ]
[ ][R][N][ ][R][ ][ ][R][ ]
[ ][ ][ ][ ][ ][ ][ ][ ][ ]

Enter a direction to move or another command
W: Up | A: Left | S: Down | D: Right
1: Shoot | 2: Debug | 3: More Options
w
Move Status: Moved!
You are stabed by the ninja, return to original point
[ ][ ][ ][I][ ][N][ ][ ][ ]   Lives: 1
[B][b][ ][ ][R][ ][ ][R][ ]   Ammo: 1
[ ][N][ ][ ][ ][N][N][ ][ ]   Invincible Turns: 0
[ ][ ][ ][ ][ ][ ][ ][ ][ ]   Radar: Disabled
[ ][R][ ][ ][R][ ][r][R][ ]
[ ][ ][N][ ][ ][ ][ ][ ][ ]
[ ][ ][ ][ ][ ][ ][ ][ ][ ]
[ ][R][N][ ][R][ ][ ][R][ ]
[S][ ][ ][ ][ ][ ][ ][ ][ ]
```

**Comment by Jester565** [comment]
The largest issue is that the whole board shouldn't be reset after every time the player dies.  I'm going to go silent and fix this but please keep putting up issues if you can.

**Comment by jtara1** [comment]
Bug - ~**spy moves into range of ninja but doesn't die**~ (fixed)

latest commit 74ddf62b2a95e0d37f32852b48a500877a0be91a

Should be solved with my later commit and comment as stated here:

>My commit in my branch 9e509b3 should allow ninjas to properly kill player now.
I think the problem was that the Sword attack(...) method was calling super.attack(...) for each of the four directions but only returning the boolean returned by the last call to super.attack(...) thus it was only attacking in one consistent direction regardless of what direction was passed as a parameter.

```
Move Status: Moved!
[ ][ ][N][ ][ ][ ][ ][r][ ]   Lives: 3
[N][R][ ][ ][R][ ][ ][R][ ]   Ammo: 1
[ ][ ][ ][ ][ ][ ][ ][ ][ ]   Invincible Turns: 0
[N][ ][I][ ][ ][ ][ ][ ][ ]   Radar: Disabled
[ ][b][N][ ][R][ ][ ][R][ ]
[S][ ][ ][ ][ ][N][ ][ ][ ]
[ ][ ][ ][N][ ][ ][ ][ ][ ]
[ ][R][ ][ ][R][ ][ ][R][ ]
[ ][ ][B][ ][ ][ ][ ][ ][ ]

Enter a direction to look in or another command
W: Up | A: Left | S: Down | D: Right
2: Debug | 3: More Options
w
Ninja Ahead!
[ ][ ][N][ ][ ][ ][ ][r][ ]   Lives: 3
[N][R][ ][ ][R][ ][ ][R][ ]   Ammo: 1
[ ][ ][ ][ ][ ][ ][ ][ ][ ]   Invincible Turns: 0
[N][ ][I][ ][ ][ ][ ][ ][ ]   Radar: Disabled
[ ][b][N][ ][R][ ][ ][R][ ]
[S][ ][ ][ ][ ][N][ ][ ][ ]
[ ][ ][ ][N][ ][ ][ ][ ][ ]
[ ][R][ ][ ][R][ ][ ][R][ ]
[ ][ ][B][ ][ ][ ][ ][ ][ ]

Enter a direction to move or another command
W: Up | A: Left | S: Down | D: Right
1: Shoot | 2: Debug | 3: More Options
w
Move Status: Moved!
[ ][N][ ][ ][ ][ ][ ][r][ ]   Lives: 3
[ ][R][ ][ ][R][ ][ ][R][ ]   Ammo: 1
[N][ ][ ][ ][ ][ ][ ][ ][ ]   Invincible Turns: 0
[ ][N][I][ ][ ][ ][ ][ ][ ]   Radar: Disabled
[S][b][ ][ ][R][ ][ ][R][ ]
[ ][ ][N][ ][ ][ ][ ][ ][ ]
[ ][ ][ ][ ][N][N][ ][ ][ ]
[ ][R][ ][ ][R][ ][ ][R][ ]
[ ][ ][B][ ][ ][ ][ ][ ][ ]

Enter a direction to look in or another command
W: Up | A: Left | S: Down | D: Right
2: Debug | 3: More Options
```

**Comment by Jester565** [comment]
Ninjas aren't killing the player in rare cases.  I'll continue to investigate.

**Comment by YaysonKwok** [comment]
Minor issue - ~~Stabbed is misspelled~~ (fixed)
```
You are stabed by the ninja, return to original point
[ ][ ][ ][ ][ ][ ][ ][ ][N]   Lives: 1
[ ][R][ ][ ][R][ ][N][R][ ]   Ammo: 1
[ ][ ][ ][ ][ ][ ][ ][B][ ]   Invincible Turns: 0
[ ][ ][ ][ ][ ][ ][ ][ ][ ]   Radar: Disabled
[ ][R][ ][ ][b][ ][ ][R][N]
[ ][ ][ ][ ][ ][N][ ][ ][ ]
[ ][ ][ ][ ][ ][ ][N][r][ ]
[ ][R][ ][ ][R][ ][I][R][ ]
[S][ ][ ][N][ ][ ][ ][ ][ ]
```

**Comment by YaysonKwok** [comment]
Bug - ~~Looked into room from north, saw a ninja?~~ (fixed)
Was testing the visibility, not sure if ninja was behind the room.

```
[*][*][*][*][*][*][*][*][*]   Lives: 2
[*][R][*][*][R][*][*][R][*]   Ammo: 1
[*][*][*][*][*][*][*][*][*]   Invincible Turns: 0
[*][*][*][*][S][*][*][*][*]   Radar: Enabled
[*][R][*][*][b][*][*][R][*]
[*][*][*][*][*][*][*][*][*]
[*][*][*][*][*][*][*][*][*]
[*][R][*][*][R][*][*][R][*]
[*][*][*][*][*][*][*][*][*]

Enter a direction to look in or another command
W: Up | A: Left | S: Down | D: Right
2: Debug | 3: More Options
s
Ninja Ahead!
[*][*][*][*][*][*][*][*][*]   Lives: 2
[*][R][*][*][R][*][*][R][*]   Ammo: 1
[*][*][*][*][*][*][*][*][*]   Invincible Turns: 0
[*][*][*][*][S][*][*][*][*]   Radar: Enabled
[*][R][*][*][b][*][*][R][*]
[*][*][*][*][*][*][*][*][*]
[*][*][*][*][*][*][*][*][*]
[*][R][*][*][R][*][*][R][*]
[*][*][*][*][*][*][*][*][*]
```

Retested - Occurs when there's a ninja in space behind the room.
```
[ ][ ][ ][r][ ][ ][ ][ ][ ]   Lives: 3
[ ][R][ ][ ][R][ ][N][b][ ]   Ammo: 1
[ ][ ][ ][N][ ][ ][N][ ][ ]   Invincible Turns: 0
[ ][N][ ][ ][N][ ][ ][I][ ]   Radar: Disabled
[ ][R][ ][ ][R][ ][ ][R][ ]
[ ][ ][ ][ ][S][ ][ ][ ][N]
[ ][ ][ ][ ][ ][ ][ ][ ][ ]
[ ][R][ ][ ][R][ ][ ][R][ ]
[ ][ ][ ][ ][ ][ ][ ][ ][ ]

Enter a direction to look in or another command
W: Up | A: Left | S: Down | D: Right
2: Debug | 3: More Options
w
Ninja Ahead!
[ ][ ][ ][r][ ][ ][ ][ ][ ]   Lives: 3
[ ][R][ ][ ][R][ ][N][b][ ]   Ammo: 1
[ ][ ][ ][N][ ][ ][N][ ][ ]   Invincible Turns: 0
[ ][N][ ][ ][N][ ][ ][I][ ]   Radar: Disabled
[ ][R][ ][ ][R][ ][ ][R][ ]
[ ][ ][ ][ ][S][ ][ ][ ][N]
[ ][ ][ ][ ][ ][ ][ ][ ][ ]
[ ][R][ ][ ][R][ ][ ][R][ ]
[ ][ ][ ][ ][ ][ ][ ][ ][ ]

Enter a direction to move or another command
W: Up | A: Left | S: Down | D: Right
1: Shoot | 2: Debug | 3: More Options
```

**Comment by jtara1** [comment]
My commit in my branch 9e509b3232bdbf64f0d17b03df5ebe198679029f should allow ninjas to properly kill player now. 

I think the problem was that the `Sword` `attack(...)` method was calling `super.attack(...)` for each of the four directions but only returning the boolean returned by the last call to `super.attack(...)` thus it was only attacking in one consistent direction regardless of what direction was passed as a parameter.

**Comment by Jester565** [comment]
Good stuff James.  I implemented that in my commit but full credit to you.   I don't want to deal with merging right now so thanks for doing it in a separate branch.

**Comment by Jester565** [comment]
I think all that should be left is fixing the resetting of the Grid after the Spy dies but tell me if there are any more bugs.

**Comment by Jester565** [comment]
I'm going to add in the feature of seeing the Ninja that killed you.  If you guys find any more bugs let me know.  James, can you do that thing where you mark this version as a stable release?  The thing you did last Tuesday.

**Comment by jtara1** [comment]
I'd rather not add the `git tag` for `v0.2` until after we hand in the code to the professor.

Unless you really want me to do so prior to your next changes.

edit: or we could just have `v0.2` now and if we need extra patches after we turn it in we could have a `v0.21` after we turn it in for the version 0.2 release.

**Comment by Jester565** [comment]
Good point.  Yeah we'll wait then.

**Comment by jrcastron** [comment]
The ninja didn't kill me...
```
w
Clear
[ ][ ][ ][ ][ ][ ][N][ ][ ]   Lives: 1
[ ][R][ ][ ][b][ ][ ][R][ ]   Ammo: 1
[ ][B][ ][ ][ ][ ][ ][N][ ]   Invincible Turns: 0
[ ][N][ ][ ][N][ ][ ][ ][ ]   Radar: Disabled
[ ][R][ ][ ][R][ ][ ][R][ ]
[ ][ ][ ][N][ ][ ][ ][ ][ ]
[ ][ ][ ][ ][I][ ][ ][ ][ ]
[ ][R][ ][ ][R][N][ ][R][ ]
[ ][ ][ ][S][ ][r][ ][ ][ ]

Enter a direction to move or another command
W: Up | A: Left | S: Down | D: Right
1: Shoot | 2: Debug | 3: More Options
w
Move Status: Moved
[ ][ ][ ][ ][ ][ ][ ][N][ ]   Lives: 1
[ ][R][ ][ ][b][ ][ ][R][ ]   Ammo: 1
[ ][B][ ][ ][N][ ][ ][ ][N]   Invincible Turns: 0
[N][ ][ ][ ][ ][ ][ ][ ][ ]   Radar: Disabled
[ ][R][ ][ ][R][ ][ ][R][ ]
[ ][ ][ ][ ][ ][ ][ ][ ][ ]
[ ][ ][ ][N][I][ ][ ][ ][ ]
[ ][R][ ][S][R][ ][N][R][ ]
[ ][ ][ ][ ][ ][r][ ][ ][ ]

Enter a direction to look in or another command
W: Up | A: Left | S: Down | D: Right
2: Debug | 3: More Options
w
Ninja Ahead!
[ ][ ][ ][ ][ ][ ][ ][N][ ]   Lives: 1
[ ][R][ ][ ][b][ ][ ][R][ ]   Ammo: 1
[ ][B][ ][ ][N][ ][ ][ ][N]   Invincible Turns: 0
[N][ ][ ][ ][ ][ ][ ][ ][ ]   Radar: Disabled
[ ][R][ ][ ][R][ ][ ][R][ ]
[ ][ ][ ][ ][ ][ ][ ][ ][ ]
[ ][ ][ ][N][I][ ][ ][ ][ ]
[ ][R][ ][S][R][ ][N][R][ ]
[ ][ ][ ][ ][ ][r][ ][ ][ ]

Enter a direction to move or another command
W: Up | A: Left | S: Down | D: Right
1: Shoot | 2: Debug | 3: More Options
w
Move Status: You can't walk onto the ninja's tile
[ ][ ][ ][ ][ ][ ][ ][N][ ]   Lives: 1
[ ][R][ ][ ][b][ ][ ][R][ ]   Ammo: 1
[ ][B][ ][ ][N][ ][ ][ ][N]   Invincible Turns: 0
[N][ ][ ][ ][ ][ ][ ][ ][ ]   Radar: Disabled
[ ][R][ ][ ][R][ ][ ][R][ ]
[ ][ ][ ][ ][ ][ ][ ][ ][ ]
[ ][ ][ ][N][I][ ][ ][ ][ ]
[ ][R][ ][S][R][ ][N][R][ ]
[ ][ ][ ][ ][ ][r][ ][ ][ ]
```

**Comment by Jester565** [comment]
That Ninja isn't supposed to kill you...  He can only kill you from his initial position.  You can't step on the Ninja either, that is an illegal move.

**Comment by jrcastron** [comment]
oh i thought they automatically killed u if u were standing next to them

**Comment by Jester565** [comment]
See issue #15

**Comment by jrcastron** [comment]
<s>what is supposed to happen when you access the main menu in the middle of the game and choose exit? cause i think right now its just starting a new game.</s> (fixed)

**Comment by Jester565** [comment]
Hmm.. I'll take care of that.  Good catch!

edit: fixed with e6f7140409e9975c3b6443e9d953332d16e87ae7

**Comment by jtara1** [comment]
Closing this issue - I think we solved all the problems (for Milestone 2), we got everyone doing thorough testing and helping fix things.

Thanks.
